py-rattler/rattler/version/with_source.py

```diff
@@ -0,0 +1,90 @@
+from __future__ import annotations
+
+from typing import Optional
+
+from rattler.version import Version
+
+
+class VersionWithSource:
+    """
+    Holds a version and the string it was created from. This is useful if
+    you want to retain the original string the version was created from.
+    This might be useful in cases where you have multiple strings that
+    are represented by the same [`Version`] but you still want to be able to
+    distinguish them.
+
+    The string `1.1` and `1.01` represent the same version. When you print
+    the parsed version though it will come out as `1.1`. You loose the
+    original representation. This class stores the original source string,
+    which can be accessed by `source` property.
+    """
+
+    def __init__(self, source: str, version: Optional[Version] = None):
```

Comment:
I think this signature should be:

```suggestion
    def __init__(self, version: Union[str, Version]):
```